### PR TITLE
PWN-7115 - New Swap. Token B list logic and search

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7115

## Description of Work

- search
- token b sorting and fetching

## Screenshots (Optional)

| Screenshot #1 |
|:-------------:|
|Screen Shot 2023-02-23 at 19 51 14|

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt [hunk 1]
```diff
@@ -7,28 +7,49 @@ import org.p2p.wallet.swap.ui.jupiter.tokens.SwapTokensContract
...
+        )
+        view?.setTokenItems(cellItems)
+    }
+
     override fun onSearchTokenQueryChanged(newQuery: String) {
+        if (newQuery.isBlank()) {
+            renderAllSwapTokensList(allTokens)
+        } else {
+            launch {
+                renderSearchTokenList(interactor.searchToken(tokenToChange, newQuery))
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt:45]: add try catch here and in attach

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensToCellItemsMapper.kt [hunk 2]
```diff
@@ -53,40 +53,57 @@ class SwapTokensToCellItemsMapper {
...
+        val otherTokens = otherTokens
             .sortedWith(otherTokensSorter)
             .map { it.toTokenFinanceCellModel() }
 
         this += sectionHeader
-        this += otherTokenItems
+        this += otherTokens
+    }
+
+    fun toSearchResultCellModels(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensToCellItemsMapper.kt:92]: maybe extract to separate mapper?

**Review by NikolaiEmelianov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt [hunk 3]
```diff
@@ -4,6 +4,7 @@ import org.p2p.wallet.home.repository.HomeLocalRepository
 import org.p2p.wallet.swap.jupiter.domain.model.SwapTokenModel
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:4]: why do you have interactor in `swap/ui` folder? 
Move it to `swap/interactor` or `swap/jupiter/interactor, or `jupiter/interactor`
Moreover, we don't have a folder like `domain`
Use `interactor` or `repository`

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt [hunk 4]
```diff
@@ -39,6 +55,37 @@ class SwapTokensInteractor(
...
+            SwapTokensListMode.TOKEN_B -> getAllAvailableTokensB()
+        }
+            .filter { filterBySymbolOrName(swapToken = it, querySymbolOrName = symbolOrName) }
+    }
+
+    private fun filterBySymbolOrName(swapToken: SwapTokenModel, querySymbolOrName: String): Boolean {
+        val (symbol, name) = when (swapToken) {
+            is SwapTokenModel.JupiterToken -> swapToken.details.tokenSymbol to swapToken.details.tokenName
+            is SwapTokenModel.UserToken -> swapToken.details.tokenSymbol to swapToken.details.tokenName
+        }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:84]: mb you will move it inside`SwapTokenModel` ?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensToCellItemsMapper.kt [hunk 5]
```diff
@@ -53,40 +53,57 @@ class SwapTokensToCellItemsMapper {
...
+        this += allOtherTokensGroup(swapTokens.filterIsInstance<SwapTokenModel.JupiterToken>())
     }
 
     private fun chosenTokenGroup(
         chosenToken: SwapTokenModel
     ): List<AnyCellItem> = buildList {
-        val sectionHeaderCell = createSectionHeader(R.string.swap_tokens_section_chosen_token)
-        val chosenTokenCell = when (chosenToken) {
+        val sectionHeader = createSectionHeader(R.string.swap_tokens_section_chosen_token)
+        val chosenToken = when (chosenToken) {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensToCellItemsMapper.kt:63]: `selectedToken`?

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt [hunk 6]
```diff
@@ -39,6 +55,37 @@ class SwapTokensInteractor(
...
+        val jupiterTokens = swapTokensRepository.getTokens().map(SwapTokenModel::JupiterToken)
+
+        val tokenA = getCurrentTokenA()
+        val tokenB = getCurrentTokenB()
+        val availableTokenBMints = swapRoutesRepository.getSwappableTokenMints(sourceTokenMint = tokenA.mintAddress)
+
+        val allTokensB = userTokens + jupiterTokens
+        return allTokensB
+            .withoutCurrentToken(tokenB.mintAddress)
+            .filter { it.mintAddress in availableTokenBMints }
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:69]: Would be better to filter only once, for example: removing current token in `filter`
You're creating a new list in `withoutCurrentToken` and in `filter` here and everywhere you are using it

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt [hunk 7]
```diff
@@ -7,28 +7,49 @@ import org.p2p.wallet.swap.ui.jupiter.tokens.SwapTokensContract
...
+            swapTokens = tokens
+        )
+        view?.setTokenItems(cellItems)
+    }
+
     override fun onSearchTokenQueryChanged(newQuery: String) {
+        if (newQuery.isBlank()) {
+            renderAllSwapTokensList(allTokens)
+        } else {
+            launch {
```
Comment (1sakov) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt:44]: maybe you'll move it inside `renderSearchTokenList`?

**Review by 1sakov — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt [hunk 8]
```diff
@@ -27,6 +28,21 @@ class SwapTokensInteractor(
 //        }
     }
 
+    suspend fun getCurrentTokenB(): SwapTokenModel {
+        return SwapTokenModel.UserToken(
+            homeLocalRepository.getUserTokens().last()
+        )
+//        return swapStateManager.getStateValue { state ->
```
Comment (Davran331122) [inline app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:35]: forgot to remove comments

**Review by Davran331122 — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt (hunk 3 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:4]: no problem, but it's not my class and not my package
SwapTokensInteractor sits in interactor package

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt (hunk 7 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/presenter/SwapTokensPresenter.kt:44]: i'll check, thanks

File: app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt (hunk 6 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/swap/ui/jupiter/tokens/interactor/SwapTokensInteractor.kt:69]: okay

**Review by 1sakov — APPROVED** [review]

**Pull request merged**